google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1alpha2/JsonStreamWriter.java

```diff
@@ -0,0 +1,359 @@
+    Matcher matcher = streamPattern.matcher(builder.streamName);
+    if (!matcher.matches()) {
+      throw new IllegalArgumentException("Invalid stream name: " + builder.streamName);
+    }
+
+    this.streamName = builder.streamName;
+    this.client = builder.client;
+    this.BQTableSchema = builder.BQTableSchema;
+    this.descriptor =
+        BQTableSchemaToProtoDescriptor.convertBQTableSchemaToProtoDescriptor(builder.BQTableSchema);
+
+    StreamWriter.Builder streamWriterBuilder;
+    if (this.client == null) {
+      streamWriterBuilder = StreamWriter.newBuilder(builder.streamName);
+    } else {
+      streamWriterBuilder = StreamWriter.newBuilder(builder.streamName, builder.client);
+    }
+    setStreamWriterSettings(
+        streamWriterBuilder,
+        builder.channelProvider,
+        builder.credentialsProvider,
+        builder.batchingSettings,
+        builder.retrySettings,
+        builder.executorProvider,
+        builder.endpoint);
+    this.streamWriter = streamWriterBuilder.build();
+  }
+
+  // For testing, use proto first.
+  // public <T extends Message> ApiFuture<AppendRowsResponse> append(JSONArray jsonArr) {
+  //   for (int i = 0; i < jsonArr.size(); i++) {
+  //     JSONObject json = jsonArr.getJSONObject(i);
+  //   }
+  //
+  // //   ProtoRows.Builder rowsBuilder = ProtoRows.newBuilder();
+  // //   Descriptors.Descriptor descriptor = null;
+  // //   for (Message protoRow : protoRows) {
+  // //     rowsBuilder.addSerializedRows(protoRow.toByteString());
+  // //   }
+  // //   AppendRowsRequest.ProtoData.Builder data = AppendRowsRequest.ProtoData.newBuilder();
+  // //
+  // data.setWriterSchema(ProtoSchemaConverter.convert(protoRows.get(0).getDescriptorForType()));
+  // //   data.setRows(rowsBuilder.build());
+  // //
+  // //   ApiFuture<AppendRowsResponse> appendResponseFuture =
+  // // this.streamWriter.append(AppendRowsRequest.newBuilder().setProtoRows(data.build()).build());
+  // //   // AppendRowsResponse appendResponse = appendResponseFuture.get();
+  // //   // if (appendResponse.hasUpdatedSchema()) {
+  // //   //
+  // //   //   // this.descriptor =
+  // // BQTableSchemaToProtoDescriptor.convertBQTableSchemaToProtoDescriptor(BQTableSchema);
+  // //   // }
+  //   // return appendResponseFuture;
+  // }
+
+  /**
+   * Updates the descriptor and BQTable schema. This function is used to support schema updates, and
+   * is called whenever AppendRowsResponse includes an updated schema. The function is synchronized
+   * since it is called through a callback (which is in another thread). If the main thread is
+   * calling append while the callback thread calls refreshAppend(), this might cause some issues.
+   *
+   * @param BQTableSchema The updated table schema.
+   */
+  private synchronized void updateDescriptor(Table.TableSchema updatedSchema)
+      throws Descriptors.DescriptorValidationException, IOException, InterruptedException {
+    if (!this.BQTableSchema.equals(updatedSchema)) {
```

Comment:
I don't think this equal check will work. One thing is that you could do a content hash and then double check on the hash. Or just assume the backend always do the write thing.

Also log some message here since this is a major operation.

Reply:
Got it, I'll just assume the backend does the right thing. I was just thinking to eliminate processing the same update schema if it gets passed twice. Will add logging!